### PR TITLE
Add optimised Rectangular Polygon

### PR DESCRIPTION
**Pull request by tocsoft**
A small improvement for when you want to draw simple rectangles

### CODE REVIEW

**Comment by tocsoft** [comment]
_'ImageSharp Fill Rectangle - As Polygon'_ is the code you would have to use now, i.e. specifying the 4 corners.

_'ImageSharp Fill Rectangle'_ is new code based on the optimised rectangle class.

Slowly getting closer and closer to that System.Drawing benchmark.

``` ini

BenchmarkDotNet=v0.10.1, OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i5-4690 CPU 3.50GHz, ProcessorCount=4
Frequency=3417967 Hz, Resolution=292.5716 ns, Timer=TSC
  [Host]     : Clr 4.0.30319.42000, 64bit RyuJIT-v4.6.1586.0 [AttachedDebugger]
  DefaultJob : Clr 4.0.30319.42000, 64bit RyuJIT-v4.6.1586.0


```
|                                   Method |      Mean |    StdDev | Scaled | Scaled-StdDev |     Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|----------------------------------------- |---------- |---------- |------- |-------------- |---------- |--------- |--------- |---------- |
|          'System.Drawing Fill Rectangle' | 2.4569 ms | 0.0236 ms |   1.00 |          0.00 | 1214.5833 | 394.2708 | 394.2708 |   5.15 MB |
|              'ImageSharp Fill Rectangle' | 3.4115 ms | 0.0629 ms |   1.39 |          0.03 |  868.7500 | 821.8750 | 821.8750 |   7.63 MB |
| 'ImageSharp Fill Rectangle - As Polygon' | 4.1909 ms | 0.0344 ms |   1.71 |          0.02 |  881.2500 | 834.3750 | 834.3750 |   7.63 MB |

File: src/ImageSharp/Drawing/Processors/DrawPathProcessor.cs [hunk 1]
```diff
@@ -87,6 +88,9 @@ public DrawPathProcessor(IPen<TColor, TPacked> pen, IPath[] paths, GraphicsOptio
         /// <inheritdoc/>
         protected override void OnApply(ImageBase<TColor, TPacked> source, Rectangle sourceRectangle)
         {
+#if DEBUG
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Drawing/Processors/DrawPathProcessor.cs]: Do we still need this?

File: tests/ImageSharp.Benchmarks/Drawing/FillRectangle.cs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+                using (Graphics graphics = Graphics.FromImage(destination))
+                {
+                    graphics.InterpolationMode = InterpolationMode.Default;
+                    graphics.SmoothingMode = SmoothingMode.AntiAlias;
+                    var pen = new Pen(Color.HotPink, 10);
+                    graphics.FillRectangle(Brushes.HotPink, new Rectangle(10, 10, 190, 140));
+                }
+
+                using (MemoryStream ms = new MemoryStream())
```
Comment (JimBobSquarePants) [inline tests/ImageSharp.Benchmarks/Drawing/FillRectangle.cs]: Don't save the bitmap in the benchmarks. We end up benchmarking the encoder also. Return the rectangle as a result BMDN works better with returns.

**Comment by tocsoft** [comment]
Re: clipper.cs 

originally I was thinking it might be better to leave clipper.cs as is because it'll make it much easier to pull in any library updates without having to re-port the code. But on the other hand porting it does give me the opportunity to add at least one optimisation idea I've had. Also means I'm going to have to peak inside the magic clipper black box and actually figure out how the thing works.

File: src/ImageSharp/Drawing/Processors/DrawPathProcessor.cs (hunk 1 above)
Comment (tocsoft) [inline src/ImageSharp/Drawing/Processors/DrawPathProcessor.cs]: my bad forgot to drop it back out after debugging something.

**Comment by JimBobSquarePants** [comment]
I think porting clipper would definitely be the best bet. It's not been updated since Jan 2014 so we should be safe for discrepancies. I think once we remove some of the compiler conditionals too and perform a cleanup we'll see places we can use numerics to improve our performance too.

It is a black box though so....



!

**Comment by tocsoft** [comment]
😆 looks like its time to go adventuring with my little kitty in tow... 🛠 🐱 🔦 😨

**Comment by tocsoft** [comment]
Benchmark results after removing the save as bmp from the benchmarks

``` ini

BenchmarkDotNet=v0.10.1, OS=Microsoft Windows NT 6.2.9200.0
Processor=Intel(R) Core(TM) i7-4770 CPU 3.40GHz, ProcessorCount=8
Frequency=3312648 Hz, Resolution=301.8733 ns, Timer=TSC
  [Host]     : Clr 4.0.30319.42000, 64bit RyuJIT-v4.6.1586.0 [AttachedDebugger]
  DefaultJob : Clr 4.0.30319.42000, 64bit RyuJIT-v4.6.1586.0


```
|                                   Method |      Mean |    StdDev |    Median | Scaled | Scaled-StdDev |    Gen 0 |    Gen 1 |    Gen 2 | Allocated |
|----------------------------------------- |---------- |---------- |---------- |------- |-------------- |--------- |--------- |--------- |---------- |
|          'System.Drawing Fill Rectangle' | 1.1379 ms | 0.1035 ms | 1.0967 ms |   1.00 |          0.00 |        - |        - |        - |     256 B |
|              'ImageSharp Fill Rectangle' | 1.2617 ms | 0.0255 ms | 1.2637 ms |   1.12 |          0.10 | 308.5938 | 304.1016 | 304.1016 |   2.59 MB |
| 'ImageSharp Fill Rectangle - As Polygon' | 2.2577 ms | 0.0569 ms | 2.2696 ms |   2.00 |          0.18 | 267.7083 | 265.1042 | 265.1042 |   2.59 MB |

**Comment by tocsoft** [comment]
created issue #47 to track clipper stuff so that can move off to a separate PR.

**Comment by JimBobSquarePants** [comment]
Happy to merge this then. Thanks!

**Pull request merged**